Fix #77726: Allow null character in regex patterns

tobil4sk opened the pull request:
Link to [#77726](https://bugs.php.net/bug.php?id=77726)

In 8b3c1a3, this was disallowed to fix [#55856](https://bugs.php.net/bug.php?id=55856), which was a security
issue caused by the /e modifier. The fix that was made was the
"Easier fix" as described in the original report.

With this fix, pattern strings are no longer treated as null terminated,
so null characters can be placed inside and matched against with regex
patterns without security problems, so there is no longer a reason to
give the error. Allowing this is consistent with the behaviour of many
other languages, including JavaScript, and thanks to PCRE2[0], it does
not require manually escaping null characters. Now that we can avoid the
error here without the cost of escaping characters, there is really no
need anymore to stray here from the conventional behaviour.

Currently, null characters are still disallowed before the first
delimiter and in the options section at the end of a regex string, but
these error messages have been updated. (This could be changed to
simply ignore them, like how whitespace is treated?)

[0] Since PCRE2, pattern strings no longer have to be null terminated,
and raw null characters match as normal.

tobil4sk commented:
Thanks for the feedback! I have retargeted the pull request.

cmb69 commented on ext/pcre/php_pcre.c:
```diff
@@ -705,13 +707,11 @@ PHPAPI pcre_cache_entry* pcre_get_compiled_regex_cache_ex(zend_string *regex, in
 	}
 
-	if (*pp == 0) {
+	if (pp == end_p) {
```
Maybe for consistency with other changes:
```suggestion
	if (pp >= end_p) {
```

tobil4sk replied:
Applied this in the latest commit.

cmb69 commented on ext/pcre/php_pcre.c:
```diff
@@ -645,28 +645,30 @@ PHPAPI pcre_cache_entry* pcre_get_compiled_regex_cache_ex(zend_string *regex, in
 	}
 
 	p = ZSTR_VAL(regex);
+	end_p = ZSTR_VAL(regex) + ZSTR_LEN(regex);
```
Since we're no longer restricted to C89, maybe declare the variable here:
```suggestion
	const char *end_p = ZSTR_VAL(regex) + ZSTR_LEN(regex);
```

tobil4sk replied:
Applied this in the latest commit as well.

tobil4sk commented on ext/pcre/php_pcre.c:
```diff
@@ -645,28 +645,30 @@ PHPAPI pcre_cache_entry* pcre_get_compiled_regex_cache_ex(zend_string *regex, in
 			zend_string_release_ex(key, 0);
 		}
-		php_error_docref(NULL,E_WARNING, "Delimiter must not be alphanumeric or backslash");
+		php_error_docref(NULL, E_WARNING, delimiter == '\0' ?
+			"Delimiter must not be a null character" :
+			"Delimiter must not be alphanumeric or backslash");
```
I have been thinking whether it might be better to simply treat NUL as whitespace here. That would avoid the need for changing or adding error messages, and instead the only change would be the removal of the messages related to the issue. Other functions, such as trim, treat it as whitespace: https://www.php.net/manual/en/function.trim.php.

cmb69 replied:
Hmm, it might be better security wise to not treat NUL bytes as whitespace here.

cmb69 replied:
Hmm, it might be better security wise to not treat NUL bytes as whitespace here.

tobil4sk replied:
Alright. I have updated the warning messages.

tobil4sk commented:
@cmb69 I have also realised that there is another warning emitted here, which got missed out. It is meant to be emitted by the `preg_replace_callback_array` function, however, this specific case is not covered by the tests and I'm not quite sure what the check is for, as using invalid delimiters seems to be handled by the other check anyway.

https://github.com/php/php-src/blob/0f96e6a0cdf9650906d090e9f305fdfbfdbcec13/ext/pcre/php_pcre.c#L2440

tobil4sk commented:
@cmb69 The only way that other warning is triggered is if one of the strings passed in the patterns argument of `preg_replace_callback_array` is a null pointer, which seems completely impossible. If it's null or an empty string then that case is already handled elsewhere. In any case, it's definitely unrelated to the delimiter. I've also looked at git blame and it's been there since the function was originally added, so it doesn't seem like it was added to fix any specific issue. `preg_replace_callback` doesn't have anything like this either.

I've removed the check, but I've added extra testing for the error cases of `preg_replace_callback_array` just to be extra safe.

devnexen commented on ext/pcre/php_pcre.c:
```diff
@@ -695,8 +695,8 @@ PHPAPI pcre_cache_entry* pcre_get_compiled_regex_cache_ex(zend_string *regex, in
-		while (*pp != 0) {
-			if (*pp == '\\' && pp[1] != 0) pp++;
+		while (pp < end_p) {
+			if (*pp == '\\' && pp + 1 < end_p) pp++;
```
nit: not sure the array access notation change is worthwhile.

cmb69 reviewed (approved):
Oh, sorry, completely forgot about this. Look good. Thanks for the work, @tobilsk!

tobil4sk commented:
@cmb69 Hi, sorry to bother again, would it be possible to have my github username next to the line in the NEWS file? Thanks.
https://github.com/php/php-src/blob/d2b7d67dacaba966a52e6bf9247a814d3eabf1a7/NEWS#L349

tobil4sk commented:
Many thanks!
